README.md
# ember-prop-types

This addon provides [React-like property management](https://facebook.github.io/react/docs/typechecking-with-proptypes.html) for components.

###### Dependencies

![Ember][ember-img]
[![NPM][npm-img]][npm-url]

###### Health

[![Travis][ci-img]][ci-url]
[![Coveralls][cov-img]][cov-url]

###### Security

[![bitHound][bithound-img]][bithound-url]

## Installation

```bash
ember install ember-prop-types
```

## Usage

**[Full Docs](http://ciena-blueplanet.github.io/ember-prop-types/)**

## Examples

### Better Components

Below is an example of a component that uses the property mixin provided by this
addon:

```js
import Ember from 'ember'
import {PropTypes} from 'ember-prop-types'

export default Ember.Component.extend({
  propTypes: {
    foo: PropTypes.string,
    bar: PropTypes.number.isRequired,
    baz: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string
    ])
  },

  getDefaultProps () {
    return {
      foo: 'This is going to be highly profitable'
    }
  }
})
```

If this mixin is being used in a class other than Component, it will need to be
mixed into the class:

```js
import Ember from 'ember'
import PropTypeMixin, {PropTypes} from 'ember-prop-types'

export default Ember.ClassName.extend(PropTypeMixin, {
  propTypes: {
    foo: PropTypes.string,
    bar: PropTypes.number.isRequired,
    baz: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string
    ])
  },

  getDefaultProps () {
    return {
      foo: 'This is going to be highly profitable'
    }
  }
})
```


#### Property Validation

The idea of *propTypes* comes from the world of React and is implemented to have
an almost identical API in the Ember world. Below is a list of possible
*propTypes* to validate against.

*   any
*   array
*   arrayOf
*   bool
*   date
*   element
*   EmberObject
*   func
*   instanceOf
*   null
*   number
*   object
*   oneOf
*   oneOfType
*   regexp
*   shape
*   string
*   symbol

#### Default Property Values

In Ember you can set default property values on a component class itself but
sometimes this bites you when you end up with a property containing an array of
selected items or a state object, where all instances of the component end up
sharing that same array or object. Uncovering this issue is not always an easy
task and so *getDefaultProps* was also implemented (thanks to the React team for
  this concept as well).

[bithound-img]: https://www.bithound.io/github/ciena-blueplanet/ember-prop-types/badges/score.svg "bitHound"
[bithound-url]: https://www.bithound.io/github/ciena-blueplanet/ember-prop-types

[ember-img]: https://img.shields.io/badge/ember-1.12.2+-orange.svg "Ember 1.12.2+"

[ci-img]: https://img.shields.io/travis/ciena-blueplanet/ember-prop-types.svg "Travis CI Build Status"
[ci-url]: https://travis-ci.org/ciena-blueplanet/ember-prop-types

[cov-img]: https://img.shields.io/coveralls/ciena-blueplanet/ember-prop-types.svg "Coveralls Code Coverage"
[cov-url]: https://coveralls.io/github/ciena-blueplanet/ember-prop-types

[npm-img]: https://img.shields.io/npm/v/ember-prop-types.svg "NPM Version"
[npm-url]: https://www.npmjs.com/package/ember-prop-types


CHANGELOG.md
# 9.0.2 (2018-10-11)
* Update PUBLISH_NODE_VERSION to 10.7.0

# 9.0.1 (2018-10-11)
* Update PUBLISH_NODE_VERSION to 10.7.0

# 9.0.0 (2018-10-04)

* Run es5 getter codemod


# 8.0.0 (2018-09-20)

Dependencies:
* **Updated** core dependencies to be compatible with ember 3.3.
* **Updated** third-party dependencies to be compatible with ember 3.3.

Addon:
* **Updated**  Ember core module import paths to match Ember.js new Modules API.
* **Replaced**  Ember.Logger with the native console.

Testing:
* **Removed** `helpers/resolver.js`, `helpers/start-app.js`, `helpers/destroy-app.js` as they are not needed with Ember new testing API.
* **Updated** `test-helper.js` to align with Ember's new testing API.
* **Added** Ember 3.3 scenario to ember-try config.


# 7.0.5 (2018-08-20)

* **Fixed** Travis API key.


# 7.0.4 (2018-07-20)

* **Fixed** Travis API key.


# 7.0.3 (2018-04-26)
* **Updated** ember-component validator to work with Ember CLI 3.2.0

# 7.0.2 (2018-03-09)
* **Updated** `ember-spread` devDependency to `^5.0.0`

# 7.0.1 (2018-03-09)
* **Updated** pull request template
* **Added** issue template
* **Updated** to `pr-bumper` version `3`
* **Updated** to node 8
* **Added** slack integration
* **Updated** `ember-frost-test` to `^4.0.1`
* **Updated** `ember-test-utils` to `^8.1.1`
* **Updated** `ember-cli-frost-blueprints` to `^5.0.2`
* **Added** `package-lock.json` file

# 7.0.0 (2018-02-19)
* **Updated** test helper to now be provided from `addon-test-support`. For example, the `createComponent` test helper previously available at `dummy/tests/helpers/ember-prop-types` is now available via `ember-prop-types/test-support/utils`

# 6.0.1 (2018-01-03)

* Update `ember-spread` devDependency to ^4.0.1

# 6.0.0 (2018-01-03)
* Update `ember-cli-google-fonts` to `^2.16.2`
* Pin `ember-cli-htmlbars-inline-precompile` to `0.3.12`
* Pin `ember-get-config` to `0.2.2`
* Move `ember-string-ishtmlsafe-polyfill` to dependency
* Remove `npm-install-security-check` dependency
* Pin `prism` Bower dependency to `1.6.0`
* Remove useLintTree ember-cli-mocha configuration
* Upgrade `ember-frost-test` to `^4.0.0`
* Install `ember-cli-frost-blueprints@^5.0.1`
* Git ignore _package-lock.json_ file until officially support Node 8
* Upgrade `ember-spread` to `^3.0.1`
  

# 5.0.3 (2017-12-27)

We experienced an issue with our release version strategy and this release should not have occurred.  It should have been part of `6.0.0`

# 5.0.2 (2017-11-22)
* **Updated** to use latest version of `ember-frost-test`
* **Updated** to use latest version of `ember-test-utils`
* **Added** `bower` package since it is no longer included by Ember CLI
* **Updated** lint to skip linting of line length in `CHANGELOG.md` file 

# 5.0.1 (2017-11-02)

* Update `ember-spread` to Ember CLI 2.12.3 compatible version as part of reversion from Ember CLI 2.16.1 to 2.12.3

# 5.0.0 (2017-11-02)

## WARNING: THIS REVERTS EMBER CLI 2.16.1 BACK TO 2.12.3

We apologize for this change. Unfortunately, due to the internal needs of our organization this became a required action.

The 2.16.1 changes are now located in the `ember-cli-2.16.1` branch and will hopefully be contained in a versioned release again in the future.

# 4.1.0 (2017-10-30)
* **Added** regexp type


# 4.0.1 (2017-10-11)
* **Updated** to Ember CLI 2.16.1


# 4.0.0 (2017-09-27)
* **Updated** to Ember CLI version 2.15.1 (uses babel 6)
* **Updated** imports to ember javascript modules syntax
* **Added** eslint rule to require usage of new modules syntax (will remove once added to ember-test-utils)
* **Updated** test helpers to now come from `addon-test-support`
* **Updated** `ember-cli-code-coverage` and configured reporters
* **Removed** `ember-string-ishtmlsafe-polyfill` - If running Ember CLI < 2.8 then add back to your consuming addon/app.

# 3.14.2 (2017-09-25)

- Adds a fallback for Object.assign for browsers that don't support it (i.e. IE).


# 3.14.1 (2017-08-02)

* Install `ember-string-ishtmlsafe-polyfill` to address #139 

# 3.14.0 (2017-08-02)
Update string validation to support Ember SafeString objects. Fixes #128.


# 3.13.4 (2017-07-10)

* Fix ember-component validator usage in `ember#lts-2-8` branch of Ember

# 3.13.3 (2017-07-10)

Upgrade ember-cli to 2.12.3


# 3.13.2 (2017-07-04)

* **Upgrade** `ember-get-config`

# 3.13.1 (2017-07-04)
* **Fix** HTMLBars linting

# 3.13.0 (2017-06-23)
* **Added** date type


# 3.12.3 (2017-06-01)
 fixed error with nameOfType if typedef does not return a function string

# 3.12.2 (2017-05-30)

- Fixed a regression in the execution order of getDefaultProps introduced in #118 


# 3.12.1 (2017-05-10)

* **Updated** the secure auth tokens in `.travis.yml`


# 3.12.0 (2017-04-24)

* **Fixed** build.

# 3.11.0 (2017-04-03)
- Log typeOf value when asserting type

# 3.10.4

* **Upgraded** `ember-get-config`.


# 3.10.3

* **Fixed** `EmberComponent` prop type to work properly with Glimmer 2.


# 3.10.2

* **Upgraded** to test against Ember 2.11.


# 3.10.1

* Attempt to fix publishing of dummy app to `gh-pages` branch.


# 3.10.0

* **Added** additional builds to CI to make sure addon works with latest versions of Ember.
* **Removed** files from npm package that aren't necessary (all of the various config files).
* **Updated** dependencies to latest versions.


# 3.9.0

* **Added** new `updatable` option to all types to flag components that shouldn't allow updates. By default all properties will allow updates so specifying `updatable: true` isn't necessary.


# 3.8.0

* **Added** `EmberComponent` prop-type for properties that come from the `{{component}}` helper.


# 3.7.1
* **Fixed** typo in dummy app.


# 3.7.0

* **Added** new alternative API for the following types:

  * `any`
  * `array`
  * `bool`
  * `element`
  * `EmberObject`
  * `func`
  * `null`
  * `number`
  * `object`
  * `string`
  * `symbol`

  **Example of new API with `any` type**

  ```js
  bar: PropTypes.any(), // Same as PropTypes.any

addon/mixins/prop-types.js
/**
 * The PropTypesMixin definition
 */
import {get, getWithDefault} from '@ember/object'
import Mixin from '@ember/object/mixin'
import {assign, merge} from '@ember/polyfills'
import {typeOf} from '@ember/utils'
import config from 'ember-get-config'

import PropTypes, {
  getDef,
  logger,
  validators
} from '../utils/prop-types'

const objectAssign = Object.assign || assign || merge

export const settings = {
  requireComponentPropTypes: getWithDefault(
    config, 'ember-prop-types.requireComponentPropTypes', false
  ),
  spreadProperty: get(config, 'ember-prop-types.spreadProperty'),
  throwErrors: getWithDefault(config, 'ember-prop-types.throwErrors', false),
  validate: get(config, 'ember-prop-types.validate'),
  validateOnUpdate: getWithDefault(config, 'ember-prop-types.validateOnUpdate', false)
}

export const helpers = {
  handleError (ctx, message) {
    logger.warn(ctx, message, settings.throwErrors)
  },

  /* eslint-disable complexity */
  validateProperty (ctx, name, def) {
    let value = get(ctx, name)

    if (value === undefined && settings.spreadProperty) {
      value = get(ctx, `${settings.spreadProperty}.${name}`)
    }

    if (value === undefined) {
      if (!def.required) {
        return
      }

      helpers.handleError(ctx, `Missing required property ${name}`)

      return
    }

    if (def.type in validators) {
      validators[def.type](ctx, name, value, def, true, settings.throwErrors)
    } else {
      helpers.handleError(ctx, `Unknown propType ${def.type}`)
    }
  },
  /* eslint-enable complexity */

  validatePropTypes (ctx) {
    const disabledForEnv = settings.validate === false
    const isProduction = !config || config.environment === 'production'
    const settingMissing = settings.validate === undefined

    if (
      disabledForEnv ||
      (settingMissing && isProduction)
    ) {
      return
    }

    const propTypesArray = [].concat(ctx.get('propTypes'))
    propTypesArray.forEach((propType) => {
      if (!propType) {
        return
      }

      Object.keys(propType).forEach(name => {
        const def = getDef(propType[name])

        if (def === undefined) {
          helpers.handleError(ctx, `propType for ${name} is unknown`)
          return
        }

        if (settings.validateOnUpdate) {
          ctx.addObserver(name, ctx, function () {
            if (def.updatable === false) {
              helpers.handleError(ctx, `${name} should not be updated`)
              return
            }

            helpers.validateProperty(this, name, def)
          })
        }

        helpers.validateProperty(ctx, name, def)
      })
    })
  }
}

export default Mixin.create({
  concatenatedProperties: ['propTypes', 'getDefaultProps'],

  getDefaultProps () {
    // Maintain compatibility for 2.x users calling this._super
    return {}
  },

  init () {
    helpers.validatePropTypes(this)

    // Note defaults is a concatenated property so this is actually an array
    // of getDefaultProps() methods all the way up the inheritance chain
    const defaults = this.getDefaultProps

    // Keep a record of any properties that came from a getDefaultProps call
    // - let child getDefaultProps functions take precedence over parents
    // - let child getDefaultProps have _access_ to parent values set by getDefaultProps
    // - don't apply defaults for properties set explicitly on the target object
    const defaultedProps = {}

    defaults.forEach((propsFunction) => {
      // If for some reason getDefaultProps() isn't a function we'll simply
      // ignore it. In the future we might want to actually assert an error
      // here.
      if (typeOf(propsFunction) !== 'function') return

      // Now let's actually call the getDefaultProps() method
      const defaultProps = propsFunction.apply(this)

      // Now iterate defaults and remove any properties that already have
      // values set on this instance _that didn't come from a previous
      // getDefaultProps_
      Object.keys(defaultProps).forEach((key) => {
        if (this.get(key) !== undefined && get(defaultedProps, key) === undefined) {
          delete defaultProps[key]
        }
      })

      // Record the properties that were defaulted
      objectAssign(defaultedProps, defaultProps)

      // Apply the defaults for this layer of the hierarchy immediately
      // @sglanzer 2017-05-29 PR #118 delayed the execution of the setProperties
      // until the end of all the getDefaultProps() calls, which meant that parent
      // layers in the getDefaultProps() hierarchy weren't available for use in
      // child functions.  This broke backwards compatibility
      // concatenated defaults with the results of this method call.
      this.setProperties(defaultProps)
    })

    this._super(...arguments)
  }
})

export {PropTypes}
